File: data-pipeline-flink/device-profile-updater/src/main/scala/org/sunbird/dp/deviceprofile/domain/DeviceProfile.scala
package org.sunbird.dp.deviceprofile.domain


import java.util

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import org.sunbird.dp.deviceprofile.task.DeviceProfileUpdaterConfig


object DeviceProfile {
  def getValueOrDefault[T](value: T, defaultValue: T): T = if (value == null) defaultValue
  else value
}

class DeviceProfile() {
  private var countryCode: String = _
  private var country: String = _
  private var stateCode: String = _
  private var state: String = _
  private var city: String = _
  private var districtCustom: String = _
  private var stateCodeCustom: String = _
  private var stateCustomName: String = _
  private var userDeclaredState: String = _
  private var userDeclaredDistrict: String = _
  private var uaspec: util.HashMap[String, String] = _
  private var devicespec: util.HashMap[String, String] = _
  private var firstAccess: Long = 0L
  private var userDeclaredOn: Long = 0L
  private var apiLastUpdatedOn: Long = 0L
  private val gson = new Gson
  private val `type` = new TypeToken[util.HashMap[String, String]]() {}.getType

  this.countryCode = ""
  this.country = ""
  this.stateCode = ""
  this.state = ""
  this.city = ""
  this.districtCustom = ""
  this.stateCodeCustom = ""
  this.stateCustomName = ""
  this.userDeclaredState = ""
  this.userDeclaredDistrict = ""
  this.uaspec = new util.HashMap[String, String]
  this.devicespec = new util.HashMap[String, String]
  this.firstAccess = 0L
  this.userDeclaredOn = 0L
  this.apiLastUpdatedOn = 0L

  def toMap(config: DeviceProfileUpdaterConfig): util.Map[String, String] = {
    val values = new util.HashMap[String, String]
    values.put(config.countryCode, DeviceProfile.getValueOrDefault(this.countryCode, ""))
    values.put(config.country, DeviceProfile.getValueOrDefault(this.country, ""))
    values.put(config.stateCode, DeviceProfile.getValueOrDefault(this.stateCode, ""))
    values.put(config.state, DeviceProfile.getValueOrDefault(this.state, ""))
    values.put(config.city, DeviceProfile.getValueOrDefault(this.city, ""))
    values.put(config.districtCustom, DeviceProfile.getValueOrDefault(this.districtCustom, ""))
    values.put(config.stateCustom, DeviceProfile.getValueOrDefault(this.stateCustomName, ""))
    values.put(config.stateCustomCode, DeviceProfile.getValueOrDefault(this.stateCodeCustom, ""))
    values.put(config.userDeclaredState, DeviceProfile.getValueOrDefault(this.userDeclaredState, ""))
    values.put(config.userDeclaredDistrict, DeviceProfile.getValueOrDefault(this.userDeclaredDistrict, ""))
    values.put(config.uaSpec, gson.toJson(DeviceProfile.getValueOrDefault(this.uaspec, new util.HashMap[String, String])))
    values.put(config.deviceSpec, gson.toJson(DeviceProfile.getValueOrDefault(this.devicespec, new util.HashMap[String, String])))
    values.put(config.firstAccess, DeviceProfile.getValueOrDefault(String.valueOf(this.firstAccess), ""))
    values.put(config.userDeclaredOn, DeviceProfile.getValueOrDefault(String.valueOf(this.userDeclaredOn), ""))
    values.put(config.apiLastUpdatedOn, DeviceProfile.getValueOrDefault(String.valueOf(this.apiLastUpdatedOn), ""))
    values
  }

  def fromMap(map: util.Map[String, String], config: DeviceProfileUpdaterConfig): DeviceProfile = {
    this.countryCode = map.getOrDefault(config.countryCode, null)
    this.country = map.getOrDefault(config.country, "")
    this.stateCode = map.getOrDefault(config.stateCode, "")
    this.state = map.getOrDefault(config.state, "")
    this.city = map.getOrDefault(config.city, "")
    this.districtCustom = map.getOrDefault(config.districtCustom, "")
    this.stateCustomName = map.getOrDefault(config.stateCustom, "")
    this.stateCodeCustom = String.valueOf(map.getOrDefault(config.stateCustomCode, ""))
    this.userDeclaredState = map.getOrDefault(config.userDeclaredState, "")
    this.userDeclaredDistrict = map.getOrDefault(config.userDeclaredDistrict, "")
    this.uaspec = gson.fromJson(map.getOrDefault(config.uaSpec, ""), `type`)
    this.devicespec = gson.fromJson(map.getOrDefault("device_spec", ""), `type`)
    this.firstAccess = map.getOrDefault("first_access", "0").asInstanceOf[Number].longValue()
    this.userDeclaredOn = map.getOrDefault(config.apiLastUpdatedOn, "0").asInstanceOf[Number].longValue()
    this.apiLastUpdatedOn = map.getOrDefault(config.apiLastUpdatedOn, "0").asInstanceOf[Number].longValue()
    this
  }
}

File: data-pipeline/device-profile-updater/src/main/java/org/ekstep/ep/samza/domain/DeviceProfile.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.ekstep.ep.samza.domain;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.Map;

public class DeviceProfile {
  private String countryCode;
  private String country;
  private String stateCode;
  private String state;
  private String city;
  private String districtCustom;
  private String stateCodeCustom;
  private String stateCustomName;
  private String userDeclaredState;
  private String userDeclaredDistrict;
  private Map<String, String> uaspec;
  private Map<String, String> devicespec;
  private Long firstAccess;
  private Long user_declared_on;
  private Long api_last_updated_on;
  private Gson gson = new Gson();
  private Type type = new TypeToken<Map<String, Object>>() {}.getType();

  public DeviceProfile() {
    this.countryCode = "";
    this.country = "";
    this.stateCode = "";
    this.state = "";
    this.city = "";
    this.districtCustom = "";
    this.stateCodeCustom = "";
    this.stateCustomName = "";
    this.userDeclaredState = "";
    this.userDeclaredDistrict = "";
    this.uaspec = new HashMap<>();
    this.devicespec = new HashMap<>();
    this.firstAccess = 0L;
    this.user_declared_on = 0L;
    this.api_last_updated_on = 0L;
  }

  public Map<String, String> toMap() {
    Map<String, String> values = new HashMap<>();
    values.put("country_code", DeviceProfile.getValueOrDefault(this.countryCode, ""));
    values.put("country", DeviceProfile.getValueOrDefault(this.country, ""));
    values.put("state_code", DeviceProfile.getValueOrDefault(this.stateCode, ""));
    values.put("state", DeviceProfile.getValueOrDefault(this.state, ""));
    values.put("city", DeviceProfile.getValueOrDefault(this.city, ""));
    values.put("district_custom", DeviceProfile.getValueOrDefault(this.districtCustom, ""));
    values.put("state_custom", DeviceProfile.getValueOrDefault(this.stateCustomName, ""));
    values.put("state_code_custom", DeviceProfile.getValueOrDefault(this.stateCodeCustom, ""));
    values.put("user_declared_state", DeviceProfile.getValueOrDefault(this.userDeclaredState, ""));
    values.put("user_declared_district", DeviceProfile.getValueOrDefault(this.userDeclaredDistrict, ""));
    values.put("uaspec", gson.toJson(DeviceProfile.getValueOrDefault(this.uaspec, new HashMap<>())));
    values.put("devicespec", gson.toJson(DeviceProfile.getValueOrDefault(this.devicespec, new HashMap<>())));
    values.put("firstaccess", DeviceProfile.getValueOrDefault(String.valueOf(this.firstAccess), ""));
    values.put("user_declared_on", DeviceProfile.getValueOrDefault(String.valueOf(this.user_declared_on), ""));
    values.put("api_last_updated_on", DeviceProfile.getValueOrDefault(String.valueOf(this.api_last_updated_on), ""));
    return values;
  }

  public DeviceProfile fromMap(Map<String, String> map) {
    this.countryCode = map.getOrDefault("country_code", "");
    this.country = map.getOrDefault("country", "");
    this.stateCode = map.getOrDefault("state_code", "");
    this.state = map.getOrDefault("state", "");
    this.city = map.getOrDefault("city", "");
    this.districtCustom = map.getOrDefault("district_custom", "");
    this.stateCustomName = map.getOrDefault("state_custom", "");
    this.stateCodeCustom = map.getOrDefault("state_code_custom", "");
    this.userDeclaredState = map.getOrDefault("user_declared_state", "");
    this.userDeclaredDistrict = map.getOrDefault("user_declared_district", "");
    this.uaspec = gson.fromJson(map.getOrDefault("uaspec", ""), type);
    this.devicespec = gson.fromJson(map.getOrDefault("device_spec", ""), type);
    this.firstAccess = Long.valueOf(map.getOrDefault("first_access", "0"));
    this.user_declared_on = Long.valueOf(map.getOrDefault("api_last_updated_on", "0"));
    this.api_last_updated_on = Long.valueOf(map.getOrDefault("api_last_updated_on", "0"));
    return this;
  }

  public static <T> T getValueOrDefault(T value, T defaultValue) {
    return value == null ? defaultValue : value;
  }
}
